### PR TITLE
Fix openstack v3 authentication

### PR DESCRIPTION
**Pull request by schaubl**
### Description

This PR allows to define the OpenStack `domain` to another value that the default `Default`.
It also adds the ability to define the `scope` of the token.

With the code for the OpenStack Identity API v3, [two new parameters were added](https://github.com/apache/libcloud/blob/v0.20.1/libcloud/common/openstack_identity.py#L925): `domain_name` and `token_scope` but it was impossible to define them to a value other than their respective default.
### Status
- done, ready for review
### Checklist (tick everything that applies)
- [x] [Code linting](http://libcloud.readthedocs.org/en/latest/development.html#code-style-guide) (required, can be done after the PR checks)
- [x] Documentation
- [x] [Tests](http://libcloud.readthedocs.org/en/latest/testing.html)
- [ ] ~~[ICLA](http://libcloud.readthedocs.org/en/latest/development.html#contributing-bigger-changes) (required for bigger changes)~~

### CODE REVIEW

**Comment by Kami** [comment]
Great, thanks!

I will do a proper review later, but from a quick glance, it looks like it would be great to add some more tests to cover all the (edge) cases.

**Comment by schaubl** [comment]
@Kami I don't know how to add tests for this fix because I need to check the body of the request but all tests seems to test the reply.

To test the body of the request I have to do it in `OpenStackIdentity_3_0_MockHttp` which is not a subclass of `unittest.TestCase` and doesn't have the `assert...` methods.

File: libcloud/common/openstack_identity.py [hunk 1]
```diff
@@ -571,14 +571,13 @@ class OpenStackIdentityConnection(ConnectionUserAndKey):
...
 
     def __init__(self, auth_url, user_id, key, tenant_name=None,
+                 domain_name=None, token_scope=None,
                  timeout=None, parent_conn=None):
         super(OpenStackIdentityConnection, self).__init__(user_id=user_id,
                                                           key=key,
                                                           url=auth_url,
                                                           timeout=timeout)
 
-        self.auth_url = auth_url
```
Comment (Kami) [inline libcloud/common/openstack_identity.py:580]: (just thinking out loud)

I assume you removed those attributes since they are duplicated aka are declared again a couple of lines below, right?

File: libcloud/common/openstack_identity.py [hunk 2]
```diff
@@ -926,8 +929,7 @@ class OpenStackIdentity_3_0_Connection(OpenStackIdentityConnection):
     ]
 
     def __init__(self, auth_url, user_id, key, tenant_name=None,
-                 domain_name='Default',
-                 token_scope=OpenStackIdentityTokenScope.PROJECT,
+                 domain_name=None, token_scope=None,
```
Comment (Kami) [inline libcloud/common/openstack_identity.py]: Any particular reason why you changed the default values? Doesn't this make it backward incompatible?

File: libcloud/common/openstack_identity.py [hunk 3]
```diff
@@ -940,32 +942,29 @@ def __init__(self, auth_url, user_id, key, tenant_name=None,
...
-                             (token_scope))
+                             (self.token_scope))
 
-        if (token_scope == OpenStackIdentityTokenScope.PROJECT and
-                (not tenant_name or not domain_name)):
+        if (self.token_scope == OpenStackIdentityTokenScope.PROJECT and
+                (not self.tenant_name or not self.domain_name)):
             raise ValueError('Must provide tenant_name and domain_name '
                              'argument')
-        elif (token_scope == OpenStackIdentityTokenScope.DOMAIN and
```
Comment (Kami) [inline libcloud/common/openstack_identity.py:962]: Any particular reason you removed this? Shouldn't we still have this check here?

File: libcloud/common/openstack_identity.py (hunk 1 above)
Comment (schaubl) [inline libcloud/common/openstack_identity.py:580]: Right

File: libcloud/common/openstack_identity.py (hunk 3 above)
Comment (schaubl) [inline libcloud/common/openstack_identity.py:962]: Because `domain_name` has a default value it cannot be None nor empty.
See:
- https://github.com/schaubl/libcloud/blob/openstack-fix_auth_v3/libcloud/common/openstack_identity.py#L592
- https://github.com/schaubl/libcloud/blob/openstack-fix_auth_v3/libcloud/common/openstack.py#L420
- https://github.com/schaubl/libcloud/blob/openstack-fix_auth_v3/libcloud/common/openstack.py#L443
- https://github.com/schaubl/libcloud/blob/openstack-fix_auth_v3/libcloud/common/openstack.py#L203

File: libcloud/common/openstack_identity.py (hunk 2 above)
Comment (schaubl) [inline libcloud/common/openstack_identity.py]: Before this PR the only [place where it is called](https://github.com/apache/libcloud/pull/744/files#diff-cb42abf0c253c7c80a89e7a9293d15a3R199) wasn't passing these two parameters so their default value was the only possible value.
With the way this parameter is passed from the driver's `__init__` to this method, it will be None if it is omitted. 
So I changed the logic from

> use the default value if it is not passed during the instanciation of OpenStackIdentity_3_0_Connection

(which is a case that cannot happens)  
to 

> use the default value if it is None

This is also why I removed the `if` below.

File: libcloud/common/openstack_identity.py (hunk 3 above)
Comment (Kami) [inline libcloud/common/openstack_identity.py:962]: Ah, thanks to for the clarification.

Also re to the comment above - I would prefer to specify default values for the arguments in the method signature instead of doing it in the method body.

Doing it in the body kinda defeats the purpose of default argument values and it also breaks things such as static code analysis (now if you want to know the default value you need to be aware of the implementation details and inspect the method body).

Would it be possible to change that back to assign default values in the method signature?

Comment (schaubl) [inline libcloud/common/openstack_identity.py:962]: Ok, I will try to do something about that.

Comment (Kami) [inline libcloud/common/openstack_identity.py:962]: Great, thanks.

**Comment by schaubl** [comment]
@Kami: Now default values are specified in the method signature.
I also added back a test case and the condition which check if `domain_name` is defined.

File: libcloud/common/openstack_identity.py [hunk 4]
```diff
@@ -940,32 +942,32 @@ def __init__(self, auth_url, user_id, key, tenant_name=None,
...
         """
         super(OpenStackIdentity_3_0_Connection,
               self).__init__(auth_url=auth_url,
                              user_id=user_id,
                              key=key,
                              tenant_name=tenant_name,
+                             domain_name=domain_name,
+                             token_scope=token_scope,
                              timeout=timeout,
                              parent_conn=parent_conn)
```
Comment (schaubl) [inline libcloud/common/openstack_identity.py:957]: I would suggest to change that code to:

``` python
def __init__(self, *args, **kwargs):
    # Docstring
    super(OpenStackIdentity_3_0_Connection, self).__init__(*args, **kwargs)
```

So the signature doesn't have to be specified twice.
But I know that `kwargs` are not appreciated, so tell me if you want me to do this change or not.

Comment (tonybaloney) [inline libcloud/common/openstack_identity.py:957]: @schaubl this is a clear case to use kwargs, its not that the project dictates not using kwargs but in the case where the keyword arguments can be listed they can.
this is fine as is.

**Comment by Kami** [comment]
@schaubl Thanks.

I will review the changes again and if everything looks good go ahead and merge them intro trunk.

**Comment by Kami** [comment]
Merged, thanks!
